### DISCUSSION TITLE
Can't connect to socket after logging out

### DISCUSSION THREAD

**Discussion by raphaelweis** [hyprwm/Hyprland#3098] [Q&A] [answered]
I run Hyprland on Archlinux, and I launch the compositor using sddm, which is reported to work pretty well according to the wiki. Once in Hyprland, everything works fine, until I exit Hyprland.

If I exit Hyprland using `hyprctl dispatch exit 1`, I get popped right back into sddm (which is expected). However, after logging back in I can no longer use the hyprctl command, I get something like this everytime:
```
$ hyprctl monitors
Couldn't connect to /tmp/hypr/9654749244117f7f150c6f2a2ce4dede6e8cbb25_1693175187/.socket.sock. (3)
```
I think I'm probably doing something wrong since I haven't seen anyone complain about this, anybody has any idea of what I could be doing wrong ?
So far I've managed to reproduce this using NixOS and Hyprland, and I can reproduce this when logging in from the TTY as well.

**Comment by memchr** [comment]
What shell are you using? It appears that it is still using the old value of env var HYPRLAND_INSTANCE_SIGNATURE from the previous hypland process.

So it's more likely that it's a configuration issue with your shell or hypland.

**Comment by raphaelweis** [comment, marked as answer]
I am running zsh as my default shell. I tried changing my default shell to bash (for which I have 0 configuration) and trying again but I got the same problem, so I'm guessing either there is something wrong in my config, or I could be missing something like an environment variable in my shell configuration.

I've also looked at the environment variable you mentionned, and it appears to stay the same value after logging out and back in, which i'm guessing is not normal behavior. I'm not really sure where to go from here, I've tried a bunch of stuff that didn't work

Here my hyprland config file if you wanna take a look
```
monitor = DP-3, 3440x1440@165, auto, 1

env = XCURSOR_SIZE, 20

exec-once = hyprctl setcursor Bibata-Modern-Ice 20

exec-once = /usr/lib/polkit-gnome/polkit-gnome-authentication-agent-1
exec-once = waybar
exec-once = wbg ~/.local/share/wallpapers/golden-state-bridge.jpg

input {
	follow_mouse = 1
	kb_layout = us
	kb_variant = intl
	sensitivity = 0
}

general {
	border_size = 2
	col.active_border = rgb(fabd2f) rgb(d79921) 45deg
	col.inactive_border = rgba(00000000)
	gaps_in = 5
	gaps_out = 10
	layout = dwindle
}

decoration {
      ... (decoration settings)
}

animations {
       ... (animation settings)
	enabled = yes
}

dwindle {
	preserve_split = true
	pseudotile = true
}

master {
	new_is_master = true
}



bind = SUPER, Q, exec, brave
... (list of bindings)
bindm = SUPER, mouse:273, resizewindow

```
Thank you for your help :)

**Reply by memchr** [reply]
How do you start zsh from hyprland, and do you use something like tmux?

**Reply by raphaelweis** [reply]
Yes I use tmux, I start it automatically with my terminal emulator, and this then launches my shell.
The binding in my config executes `alacritty -e tmux`

Once I get home I'll try not using tmux and see if it changes something

**Comment by memchr** [comment]
> Yes I use tmux, I start it automatically with my terminal emulator, and this then launches my shell. The binding in my config executes `alacritty -e tmux`
> 
> Once I get home I'll try not using tmux and see if it changes something

This explains why you get the old variables, because the tmux uses the first instance as a server, which keeps all the environment variables you had when it was created.

To work around this, you can create a script that hyprland runs on login to update its Wayland-related variables.

for example, save the following as ~/.config/tmux/import_env
```bash
#!/usr/bin/env bash

set -e

[[ -n $HYPRLAND_DEBUG_CONF ]] && exit 0

_envs=(
	WAYLAND_DISPLAY
	DISPLAY
	XCURSOR_SIZE
	USERNAME
	XDG_BACKEND
	XDG_CURRENT_DESKTOP
	XDG_SESSION_TYPE
	XDG_SESSION_ID
	XDG_SESSION_CLASS
	XDG_SESSION_DESKTOP
	XDG_SEAT
	XDG_VTNR
	# sway
	SWAYSOCK
	# hyprland
	HYPRLAND_CMD
	HYPRLAND_INSTANCE_SIGNATURE
	# toolkit
	_JAVA_AWT_WM_NONREPARENTING
	QT_QPA_PLATFORM
	QT_WAYLAND_DISABLE_WINDOWDECORATION
	GRIM_DEFAULT_DIR
	# ssh
	SSH_AUTH_SOCK
)

for v in "${_envs[@]}"; do
	if [[ -n ${!v} ]]; then
		tmux setenv -g "$v" "${!v}"
	fi
done
```
Then, in hyprland.conf. exec it once
```
exec-once = ~/.config/tmux/import_env
```

**Reply by khaneliman** [reply]
I'm gonna have to try this out... I experience the issue reported but never bothered investigating. Never thought of tmux fucking up my system like that lol

**Reply by raphaelweis** [reply]
Let me know if you encounter problems, for me the script worked on the first try

**Reply by memchr** [reply]
I also forgot to mention that you may also need to import the correct variables into the dbus and systemd user sessions, otherwise the programs that use dbus activation or are started by systemd user manager won't have the correct WAYLAND_DISPLAY, XDG_SESSION_ID, etc.

e.g.

```bash
#!/usr/bin/env bash

set -e

[[ -n $HYPRLAND_DEBUG_CONF ]] && exit 0

_envs=(
	# display
	WAYLAND_DISPLAY
	DISPLAY
	# xdg
	USERNAME
	XDG_BACKEND
	XDG_CURRENT_DESKTOP
	XDG_SESSION_TYPE
	XDG_SESSION_ID
	XDG_SESSION_CLASS
	XDG_SESSION_DESKTOP
	XDG_SEAT
	XDG_VTNR
	# hyprland
	HYPRLAND_CMD
	HYPRLAND_INSTANCE_SIGNATURE
	# sway
	SWAYSOCK
	# ssh
	SSH_AUTH_SOCK
)
dbus-update-activation-environment --systemd "${_envs[@]}"

```

It can be added to hyprland.conf in the same manner
```conf
exec-once = ~/.config/hypr/import_env
```

**Reply by memchr** [reply]
I've merged the two scripts into a single once, you can use a subcommand to instruct where to import the environment variable.

[import_env.sh](https://gist.github.com/memchr/a40f9fdb95f9e263fc7a3f6a784b3feb)

e.g.

```conf
exec-once = $HOME/.local/lib/import_env tmux
exec-once = $HOME/.local/lib/import_env system
```

**Reply by jqcorreia** [reply]
@memchr you're the man. I was having exactly the same problem but when developing my launcher shell that I wanted to connect to hyprland socket. I suspected tmux was screwing my env vars but the dbus tip solves a lot of other issues I was having with my laptop, as it pops up to SDDM the second I close the lid. Thanks dude, really!

**Comment by raphaelweis** [comment]
This works flawlessly.

Thank you very much sir I really appreciate the time you took to help me.

**Reply by raphaelweis** [reply]
I'm sorry I selected the wrong thing as answer I think but all the credit goes to you

**Reply by memchr** [reply]
Thanks for the credit, I appreciate it, but there's no need for such formality.

**Reply by TwelveNight** [reply]
emm,it`s not very help for me,I use tmux too,I only can new a tmux session to connect to the sock
